src/esn.py

```diff
@@ -209,3 +425,209 @@ def update_weights(self, xt: Array, ut: Array, yt_hat: Array,
         :param optimizer: optimizer object, e.g. linear regression.
         """
         self.w_out = self._ridge_regression(xt, ut, yt_hat, alpha=alpha)
+
+
+# Auto-conceptor parts: to factorize (and combined with RCF) after debugged
+
+    def _auto(self, ut: Array, learning_rate: float, aperture: float, x_init: Optional[Array] = None, 
+              y_init: Optional[Array] = None, C_init: Optional[Array] = None,
+             collect_states: bool = False, recall: bool = False)\
+            -> Tuple[Array, ...]:
+        """
+        Forward pass for cueing, collects all reservoir states, outputs, and conceptors.
+
+        :param ut: (T, K)
+        :param x_init: (N, 1)
+        :param y_init: (L, 1)
+        :param C_init: (N, N)
+        :param learning_rate: float, learning rate for incremental adapatation of C
+        :param aperture: float
+        :return xt: (T, N)
+        :return yt: (T, L)
+        :return Ct: (T*N, N)
+
+        If collect_states == False, only returns the cued conceptor and the last reservoir state vector.
+        :return x: (N, 1)
+        :return y: (L, 1)
+        :return C: (N, N)
+        """
+
+        _, N, _, L = self.get_sizes()
+        T = ut.shape[0]
+        if collect_states:
+            xt = []
+            yt = []
+            Ct = []
+        # initial reservoir state (default: zero)
+        x = np.zeros((N, 1)) if x_init is None else x_init.copy()
+        y = np.zeros((L, 1)) if y_init is None else y_init.copy()
+        C = np.zeros((N, N)) if C_init is None else C_init.copy()
+        loaded_input = np.zeros((N, 1))
+        # time loop
+        for t in range(T):
+            u = ut[t:t+1, :].T
+            # conceptor adaptation
+            C = C + learning_rate*( np.dot(x - np.dot(C, x), x.T) - (aperture ** (-2))*C )
+            if recall:
+                loaded_input = np.dot(self.d, x)
+            # state update (with or without feedback)
+            x = np.dot(self.w_in, u) + np.dot(self.w, x) + self.b + loaded_input
+
+            if self.feedback:
+                x += np.dot(self.w_fb, y)
+            x = np.tanh(x)
+            if recall:
+                x = np.dot(C, x)
+            # compute output
+            if self.skip_connections:
+                y = np.dot(self.w_out, np.concatenate([x, u], axis=0))
+            else:
+                y = np.dot(self.w_out, x)
+            
+            if collect_states:
+                xt.append(x)
+                yt.append(y)
+                Ct.append(C)
+
+        if collect_states:
+            xt = np.concatenate(xt, axis=1).T
+            yt = np.concatenate(yt, axis=1).T
+            return xt, yt, Ct
+        else:
+            return x, y, C
+
+    def cueing(self, ut: Array, learning_rate: float, aperture: float,
```

Comment:
`cueing` and `forward_cueing` only differ in whether or not to collect the states right? then `forward_cueing` and `harvest_cueing` (or smith like that) would be a better name to understand it more easily
```suggestion
    def harvest_cueing(self, ut: Array, learning_rate: float, aperture: float,
```

Reply:
Okay, I changed it.